The job **`x86_64-gnu-distcheck`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25642629771/job/75265623631) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/75265623631) [(plain)](https://github.com/rust-lang/rust/commit/a6ba4ae32a0e33a3b990081566f0736e4bd50bdd/checks/75265623631/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  failed to read root of directory source: /checkout/obj/build/tmp/tarball/rustc/src/image/vendor

Caused by:
  No such file or directory (os error 2)
failed to run: /tmp/distcheck/distcheck-rustc-src/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --jobs=default --manifest-path /tmp/distcheck/distcheck-rustc-src/src/bootstrap/Cargo.toml -Zroot-dir=/tmp/distcheck/distcheck-rustc-src -Zwarnings --frozen
Build completed unsuccessfully in 0:00:12
make: *** [Makefile:49: check] Error 1
Bootstrap failed while executing `test distcheck`
Command `make check [workdir=/tmp/distcheck/distcheck-rustc-src]` failed with exit code 2
Created at: src/bootstrap/src/core/build_steps/test.rs:3501:5
Executed at: src/bootstrap/src/core/build_steps/test.rs:3507:10

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: bootstrap::core::build_steps::test::distcheck_plain_source_tarball
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:3507:10
   5: <bootstrap::core::build_steps::test::Distcheck as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:3470:9
   6: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Distcheck>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1595:36
   7: <bootstrap::core::build_steps::test::Distcheck as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:3453:21
   8: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   9: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
  10: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1138:9
  11: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1117:14
  12: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  13: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  14: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:250:5
  15: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/sys/backtrace.rs:166:18
  16: std::rt::lang_start::<()>::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:206:18
  17: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:287:21
  18: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
  19: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:544:19
  20: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panic.rs:359:14
  21: std::rt::lang_start_internal::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:175:24
  22: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
---
  29: __libc_start_main
  30: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:08:18
  local time: Sun May 10 23:35:46 UTC 2026
  network time: Sun, 10 May 2026 23:35:46 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/src/bootstrap/src/core/build_steps/dist.rs
+++ b/src/bootstrap/src/core/build_steps/dist.rs
@@ -1216,7 +1216,7 @@ impl Step for Src {
             sync_args: vec![],
             versioned_dirs: true,
             root_dir: dst_src.clone(),
-            output_dir: dst_src.clone(),
+            output_dir: None,
             only_library_workspace: true,
         });
 
@@ -1389,7 +1389,7 @@ fn prepare_source_tarball<'a>(
             sync_args: pkgs_for_pgo_training.collect(),
             versioned_dirs: true,
             root_dir: plain_dst_src.into(),
-            output_dir: plain_dst_src.into(),
+            output_dir: None,
             only_library_workspace: false,
         });
 

--- a/src/bootstrap/src/core/build_steps/run.rs
+++ b/src/bootstrap/src/core/build_steps/run.rs
@@ -275,7 +275,7 @@ impl Step for GenerateCopyright {
                 sync_args: Vec::new(),
                 versioned_dirs: true,
                 root_dir: builder.src.clone(),
-                output_dir: cache_dir.clone(),
+                output_dir: Some(cache_dir.clone()),
                 only_library_workspace: false,
             });
             cache_dir.join(VENDOR_DIR)

--- a/src/bootstrap/src/core/build_steps/vendor.rs
+++ b/src/bootstrap/src/core/build_steps/vendor.rs
@@ -46,10 +46,13 @@ pub(crate) struct Vendor {
     /// Determines whether vendored dependencies use versioned directories.
     pub(crate) versioned_dirs: bool,
     /// The root directory of the source code.
+    ///
+    /// Vendored dependencies will be stored in <root_dir>/vendor and
+    /// <root_dir>/library/vendor unless overridden by `output_dir`.
     pub(crate) root_dir: PathBuf,
     /// The root directory for storing vendored dependencies in <output_dir>/vendor
     /// and <output_dir>/library/vendor.
-    pub(crate) output_dir: PathBuf,
+    pub(crate) output_dir: Option<PathBuf>,
     /// Only vendor crates necessary by the library workspace.
     pub(crate) only_library_workspace: bool,
 }
@@ -71,7 +74,7 @@ impl Step for Vendor {
             sync_args: run.builder.config.cmd.vendor_sync_args(),
             versioned_dirs: run.builder.config.cmd.vendor_versioned_dirs(),
             root_dir: run.builder.src.clone(),
-            output_dir: run.builder.src.clone(),
+            output_dir: None,
             only_library_workspace: false,
         });
     }
@@ -81,7 +84,7 @@ impl Step for Vendor {
     /// This function runs `cargo vendor` and ensures all required submodules
     /// are initialized before vendoring begins.
     fn run(self, builder: &Builder<'_>) -> Self::Output {
-        let _guard = builder.group(&format!("Vendoring sources to {:?}", self.output_dir));
+        let _guard = builder.group(&format!("Vendoring sources to {:?}", self.root_dir));
 
         let config = if self.only_library_workspace {
             String::new()
@@ -116,7 +119,11 @@ impl Step for Vendor {
             cmd.env("RUSTC_BOOTSTRAP", "1");
             cmd.env("RUSTC", &builder.initial_rustc);
 
-            cmd.current_dir(&self.root_dir).arg(self.output_dir.join(VENDOR_DIR));
+            cmd.current_dir(&self.root_dir);
+            match &self.output_dir {
+                None => cmd.arg(VENDOR_DIR),
+                Some(output_dir) => cmd.arg(output_dir.join(VENDOR_DIR)),
+            };
 
             cmd.run_capture_stdout(builder).stdout()
         };
@@ -133,8 +140,11 @@ impl Step for Vendor {
         cmd.env("RUSTC_BOOTSTRAP", "1");
         cmd.env("RUSTC", &builder.initial_rustc);
 
-        cmd.current_dir(self.root_dir.join("library"))
-            .arg(self.output_dir.join("library").join(VENDOR_DIR));
+        cmd.current_dir(self.root_dir.join("library"));
+        match &self.output_dir {
+            None => cmd.arg(VENDOR_DIR),
+            Some(output_dir) => cmd.arg(output_dir.join("library").join(VENDOR_DIR)),
+        };
 
         let config_library = cmd.run_capture_stdout(builder).stdout();
 

